### PR TITLE
Implemented http fetch (partial #4576)

### PR DESCRIPTION
**Pull request by KiChjang**
This is a work-in-progress for the implementation of HTTP fetch. Currently, it does not compile, as I haven't figured out how to solve the issues regarding borrows and lifetimes.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5824)

### CODE REVIEW

**Comment by jdm** [comment]
\o/

**Comment by jdm** [comment]
@KiChjang Would you like feedback on anything specifically?

**Comment by Manishearth** [comment]
:dancers:

**Comment by KiChjang** [comment]
Yes - I've been trying to destructure and extract the `cache` from the Request object, and I've used `self.cache.unwrap()` and `if let Some(ref cache) = self.cache` to do so, but both of these attempts tell me that I cannot move out of borrowed content.

File: components/net/fetch/request.rs [hunk 1]
```diff
@@ -71,6 +84,7 @@ pub struct Request {
     pub preserve_content_codings: bool,
     // pub client: GlobalRef, // XXXManishearth copy over only the relevant fields of the global scope,
                               // not the entire scope to avoid the libscript dependency
+    pub is_service_worker_global_scope: Option<bool>,
```
Comment (Manishearth) [inline components/net/fetch/request.rs]: What's the spec link for this one?

**Comment by jdm** [comment]
Usually you want `self.cache.as_ref().unwrap()` for the first, but that should be equivalent to the `if let` form you gave. What is the error you get from doing that?

File: components/net/fetch/request.rs (hunk 1 above)
Comment (KiChjang) [inline components/net/fetch/request.rs]: This isn't actually from the spec, it's a workaround for the client (refer to the note you left on the line above).

Comment (Manishearth) [inline components/net/fetch/request.rs]: oh, I see.

I'll look through this sometime tomorrow I guess (got some changes to review in humpty_dumpty first)

**Comment by KiChjang** [comment]
Sorry, somehow git messed up my changes, and one of my commits are gone. I'll have to rewrite some code and push it before I can answer.

**Comment by jdm** [comment]
@KiChjang What feedback are you looking for here? Is this a PR that you would like to merge after review?

**Comment by KiChjang** [comment]
Yes, since all the build errors are fixed, I'd like to merge this PR after reviews.

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes +S-needs-rebase 

---

Reviewed files:
- components/net/fetch/request.rs @ r2

---

<sup>**[components/net/fetch/request.rs, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobQLEXc7kkUHrtx4Cw)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L102)):</sup>
usize

---

<sup>**[components/net/fetch/request.rs, line 108 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobQPpbbW6rJmIQ2Klp)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L108)):</sup>
Why is this an Option?

---

<sup>**[components/net/fetch/request.rs, line 168 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Job_CEfqLKwB2UELOAu)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L168)):</sup>
_authentication_fetch_flag and _cors_flag are used, so they shouldn't have the _ prefix.

---

<sup>**[components/net/fetch/request.rs, line 186 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobX_w_0tJbNUDNtLXy)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L186)):</sup>
Let's call this method_mismatch.

---

<sup>**[components/net/fetch/request.rs, line 187 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobXbrUws6ett1GMyIG)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L187)):</sup>
Let's call this header_mismatch.

---

<sup>**[components/net/fetch/request.rs, line 207 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobR4oMyupmmmS0pZ5q)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L207)):</sup>
nit: no need for ()

---

<sup>**[components/net/fetch/request.rs, line 209 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobYbpiMdAEWn25oCDE)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L209)):</sup>
This `if let` seems a bit silly - let's just do:
```rust
let preflight_result = self.preflight_fetch();
if preflight_result.response_type == ResponseType::Error {
    return Response::network_error();
}
```

---

<sup>**[components/net/fetch/request.rs, line 225 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Job_3xGPMltWejfpbDd)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L225)):</sup>
&& not ||

---

<sup>**[components/net/fetch/request.rs, line 231 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobaPI_zjpfE8Td2ntg)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L231)):</sup>
```rust
let fetch_result = self.http_network_or_cache_fetch(credentials, _authentication_fetch_flag);
if _cors_flag && self.cors_check(&fetch_result).is_err() {
    return Response::network_error();
}
response = Some(fetch_result);
```

---

<sup>**[components/net/fetch/request.rs, line 238 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Joba86NBPKPGlKLmZ8B)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L238)):</sup>
There should always be a response by now, so I would prefer `let mut response = response.unwrap();` and unindenting the rest of this code.

---

<sup>**[components/net/fetch/request.rs, line 251 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Jobb83NhFktVgVwpmi9)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L251)):</sup>
This is step 2.

---

<sup>**[components/net/fetch/request.rs, line 255 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobbN8ue5VQOjKJcTLx)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L255)):</sup>
Step 3 says return response (according to https://fetch.spec.whatwg.org/#concept-header-parse null means the header is not present). I don't think hyper exposes "failure to parse" to us, so that's worth a FIXME.

---

<sup>**[components/net/fetch/request.rs, line 263 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobbynxBH-GLZ3NFAlW)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L263)):</sup>
```rust
let locationUrl = match locationUrl {
    Ok(locationUrl) => locationUrl,
    Err(_) => return Response::network_error(),
};
```

---

<sup>**[components/net/fetch/request.rs, line 273 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobcLGP7gesRvfyk3TI)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L273)):</sup>
This is not step 9 or step 10.

---

<sup>**[components/net/fetch/request.rs, line 279 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Jobcpy6cNd6ht6sxA9N)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L279)):</sup>
Link to the rust-url issue here, please.

---

<sup>**[components/net/fetch/request.rs, line 283 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Jobd0XvH-i2OagHOo8k)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L283)):</sup>
!locationUrl.username().unwrap_or("").is_empty()

---

<sup>**[components/net/fetch/request.rs, line 286 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Jobd_Q5PcwqkrjHPDVI)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L286)):</sup>
Substep 3 is missing here.

---

<sup>**[components/net/fetch/request.rs, line 287 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobdeYUrJ0kOaLqVdaC)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L287)):</sup>
Substep 4

---

<sup>**[components/net/fetch/request.rs, line 289 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobdgmL6u-_0zAIn01X)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L289)):</sup>
Substep 5

---

<sup>**[components/net/fetch/request.rs, line 325 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Jobeurlo9sScG9sDRGw)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L325)):</sup>
This isn't specified as part of step 6; where does this come from?

---

<sup>**[components/net/fetch/request.rs, line 337 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobfDt872C77UeFk4Xo)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L337)):</sup>
No step 7 in the spec.

---

<sup>**[components/net/fetch/request.rs, line 358 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobfKZUwFknL3aGZ_dt)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L358)):</sup>
http://doc.servo.org/hyper/header/enum.IfRange.html

---

<sup>**[components/net/fetch/request.rs, line 365 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobfWDdbhc4Bu6DLG35)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L365)):</sup>
http://doc.servo.org/hyper/header/struct.AcceptLanguage.html and http://doc.servo.org/hyper/header/struct.ContentLanguage.html

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by KiChjang** [comment]
<sup>**[components/net/fetch/request.rs, line 108 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobQPpbbW6rJmIQ2Klp)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L108)):</sup>
From what I read, the global object can be 1 of 3 values: Window, WorkerGlobalScope or ServiceWorkerGlobalScope. I used that to encapsulate these 3 cases, but I guess a simple boolean would do since the global object is only checked once here in the whole Fetch spec.

---

<sup>**[components/net/fetch/request.rs, line 325 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-Jobeurlo9sScG9sDRGw)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L325)):</sup>
Oh crap. There are a couple of instances where I scrolled to nowhere in the spec and I got lost scanning the steps that I needed to partake. This is one of them.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by jdm** [comment]
-S-awaiting-review -S-needs-rebase +S-needs-code-changes 
Nearly there! This is looking really good.

---

Reviewed files:
- components/net/fetch/request.rs @ r3

---

<sup>**[components/net/fetch/request.rs, line 255 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobbN8ue5VQOjKJcTLx)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L255)):</sup>
This match still isn't right - according to step 3, when there is no header present we should return the response. Additionally, we shouldn't be comparing against the value "null" - if there is a header present, we should go directly to step 5.

---

<sup>**[components/net/fetch/request.rs, line 258 \[r3\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpEbiOWkX0Z_n-BrkOK)** ([raw file](https://github.com/servo/servo/blob/94fed7f2487e0afc39acd328d22cb80067079456/components/net/fetch/request.rs#L258)):</sup>
```rust
if ... {
    return Response::network_error();
}
```

---

<sup>**[components/net/fetch/request.rs, line 263 \[r3\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpEafOW3vDTA0AhLMzw)** ([raw file](https://github.com/servo/servo/blob/94fed7f2487e0afc39acd328d22cb80067079456/components/net/fetch/request.rs#L263)):</sup>
While true, what we really mean is "Hyper does not expose the presence of headers that failed to parse."

---

<sup>**[components/net/fetch/request.rs, line 358 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobfKZUwFknL3aGZ_dt)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L358)):</sup>
This comment can be removed now.

---

<sup>**[components/net/fetch/request.rs, line 365 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobfWDdbhc4Bu6DLG35)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L365)):</sup>
This comment can be removed now.

---

<sup>**[components/net/fetch/request.rs, line 375 \[r3\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpEbQVRjJuPueIRUhRc)** ([raw file](https://github.com/servo/servo/blob/94fed7f2487e0afc39acd328d22cb80067079456/components/net/fetch/request.rs#L375)):</sup>
Let's structure this method as:
```rust
if h.is::<ContentType>() {
    ...
} else {
    h.is::<Accept>() || h.is::<AcceptLanguage>() || h.is::<ContentLanguage>()
}
```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by KiChjang** [comment]
@seanmonstar

**Comment by seanmonstar** [comment]
Review status: 0 of 1 files reviewed, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/net/fetch/request.rs, line 146 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpYz09LHr--jbZY6K_w)** ([raw file](https://github.com/servo/servo/blob/ec455f36ed9c5e391110ae38ff61593a5547a37d/components/net/fetch/request.rs#L146)):</sup>
Couldn't find anything at this link...

---

<sup>**[components/net/fetch/request.rs, line 148 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpYz4zQfLOOQBcBGAtb)** ([raw file](https://github.com/servo/servo/blob/ec455f36ed9c5e391110ae38ff61593a5547a37d/components/net/fetch/request.rs#L148)):</sup>
This method always errors?

---

<sup>**[components/net/fetch/request.rs, line 267 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5824#-JobbN8ue5VQOjKJcTLx)** ([raw file](https://github.com/servo/servo/blob/e949b1495ef9c92a4ec787035c3eb483083787c0/components/net/fetch/request.rs#L267)):</sup>
One way to do so is to check `headers.has::<Foo>()`. If it has it, but `.get::<Foo>()` returns `None`, that means the `Foo::parse_header` method failed.

Changing the signature of `get()` to return `Option<Result<H, Error>>` seems gross...

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by KiChjang** [comment]
<sup>**[components/net/fetch/request.rs, line 146 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpYz09LHr--jbZY6K_w)** ([raw file](https://github.com/servo/servo/blob/ec455f36ed9c5e391110ae38ff61593a5547a37d/components/net/fetch/request.rs#L146)):</sup>
My bad. Let me fix it up on a new commit

---

<sup>**[components/net/fetch/request.rs, line 148 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpYz4zQfLOOQBcBGAtb)** ([raw file](https://github.com/servo/servo/blob/ec455f36ed9c5e391110ae38ff61593a5547a37d/components/net/fetch/request.rs#L148)):</sup>
This and subsequent functions like http_or_cache_fetch all return network_errors because they're unimplemented yet. I plan to only finish up http_fetch in this PR and work on the others later.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by seanmonstar** [comment]
Review status: 0 of 1 files reviewed, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/net/fetch/request.rs, line 148 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpYz4zQfLOOQBcBGAtb)** ([raw file](https://github.com/servo/servo/blob/ec455f36ed9c5e391110ae38ff61593a5547a37d/components/net/fetch/request.rs#L148)):</sup>
I don't know what servo's convetions are with this, so if I'm wrong, ignore this, but I personally would use `unimplemented!()`, to catch that the those methods shouldn't be called yet. Or perhaps they can be called because it's dependent on the DOM calling things, and you want to just return `network_error()`...

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by jdm** [comment]
Review status: all files reviewed, 3 unresolved discussions, all commit checks successful.
Reviewed files:
- components/net/fetch/request.rs @ r7

---

<sup>**[components/net/fetch/request.rs, line 148 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpYz4zQfLOOQBcBGAtb)** ([raw file](https://github.com/servo/servo/blob/ec455f36ed9c5e391110ae38ff61593a5547a37d/components/net/fetch/request.rs#L148)):</sup>
A TODO here wouldn't go amiss, I guess.

---

<sup>**[components/net/fetch/request.rs, line 267 \[r7\]](https://reviewable.io:443/reviews/servo/servo/5824#-JpbI9HzMjJelIEAqpIP)** ([raw file](https://github.com/servo/servo/blob/1a5057cac00168a4c81a06bf18160432b0823f81/components/net/fetch/request.rs#L267)):</sup>
I don't think we need any special handling here - let's just replace this match with
```rust
if location.is_none() {
    return Response::network_error();
}
```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-squash 

---

Review status: all files reviewed, 2 unresolved discussions, all commit checks successful.
Reviewed files:
- components/net/fetch/request.rs @ r8

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5824)

**Comment by jdm** [comment]
@bors-servo: retry
* #5928

**Pull request merged**